engine/src/flutter/shell/platform/linux/fl_renderer.cc

```diff
@@ -564,6 +641,41 @@ gboolean fl_renderer_present_layers(FlRenderer* self,
   return TRUE;
 }
 
+struct _FlRendererPresentLayersData {
+  FlRenderer* self;
+  FlutterViewId view_id;
+  const FlutterLayer** layers;
+  size_t layers_count;
+  gboolean* res;
+};
+
+static void fl_renderer_present_layers_trampoline(gpointer trampoline_data) {
+  _FlRendererPresentLayersData* data =
+      reinterpret_cast<_FlRendererPresentLayersData*>(trampoline_data);
+  fl_renderer_make_current(data->self);
+  *data->res = fl_renderer_present_layers_on_main_thread(
+      data->self, data->view_id, data->layers, data->layers_count);
+  fl_renderer_clear_current(data->self);
+}
+
+gboolean fl_renderer_present_layers(FlRenderer* self,
+                                    FlutterViewId view_id,
+                                    const FlutterLayer** layers,
+                                    size_t layers_count) {
+  gboolean res = FALSE;
+  _FlRendererPresentLayersData data = {
+      .self = self,
+      .view_id = view_id,
+      .layers = layers,
+      .layers_count = layers_count,
+      .res = &res,
+  };
+  fl_renderer_clear_current(self);
```

Comment:
This will be called from the render thread, which may not allowed from GTK.

Reply:
I think `gdk_gl_context_make_current` and `gdk_gl_context_clear_current` are an exception here. I went through Gtk 3 source code and for realized contexts `gdk_gl_context_make_current` basically just calls `glXMakeContextCurrent` (X11) and `eglMakeCurrent` (Wayland), both of which have defined semantics when called from multiple threads (i.e. it is allowed, but can only be tied to one thread at a time). There doesn't seem to be any other Gtk state mutation performed from these methods.

There is also this issue https://gitlab.gnome.org/GNOME/gtk/-/issues/1315, where the implication is that it should work (and the windows behavior where it doesn't is a bug in libepoxy that is being fixed).